crates/cxx-qt-gen/src/writer/rust/mod.rs

```diff
@@ -90,8 +90,15 @@ fn cxx_qt_common_blocks(qobject: &GeneratedRustQObject) -> Vec<TokenStream> {
 
     vec![
         quote! {
+            // CxxQtThread is safe to be sent across threads and handles
+            // locking and checks with the original QObject to prevent issues
             unsafe impl Send for #cxx_qt_thread_ident {}
         },
+        quote! {
+            // CxxQtThread is safe to use as a reference in parallel from multiple
+            // places as it protects the queue call and the closure with mutexes
+            unsafe impl Sync for #cxx_qt_thread_ident {}
```

Comment:
Need to check that this does work with `async` blocks

Reply:
So this allows it to be used in async blocks, but still doesn't quite work with my usecase due to issues with higher order lifetimes etc. Something to do with the way `tokio::join!(my_future)`  works

Reply:
You can see this fun here https://gitlab.com/ahayzen/home-dash-alpha-g1/-/merge_requests/3

The error i get is

```rust
error: higher-ranked lifetime error
  --> src/dashboard/src/backend/weathercontroller.rs:27:13
   |
27 | /             tokio::spawn(async move {
28 | |                 // Connect to the server
29 | |                 let (ws_stream, _) = connect_async(client_ws_url()).await.unwrap();
30 | |                 let (mut write, read) = ws_stream.split();
...  |
66 | |                 tokio::join!(read_future, write_future);
67 | |             });
   | |______________^
   |
   = note: could not prove `[async block@src/dashboard/src/backend/weathercontroller.rs:27:26: 67:14]: std::marker::Send`
```

Which suggests something in the async block it can't prove is `Send`, i might need to add a where clause somewhere.

But what is interesting is that changing `tokio::spawn!(read_future, write_future)` into `read_future.await; write_future.await;` does work. And adding `pin_mut!(read_future, write_future)` doesn't help. And if the `qt_thread.queue(...)` part is removed then it works. So some combination of `tokio::spawn` and using `qt_thread` breaks things with these `async` rather than `async move` blocks.

Reply:
@LeonMatthesKDAB I think we just merge this as-is as it still helps for reference and async blocks it just didn't totally solve the usecase i had before.

Reply:
Agreed